### PR TITLE
Some changes I did to make the forum a bit "better".

### PR DESCRIPTION
**Pull request by Blackskyliner**
Hi there,

currently I am developing an fork as I needed some features for my own needs, now I would like to let others have their benefits from it.

I'm kinda new to this git thingy, so I will just PullRequest all my commits in the first place, if I need to change smth. in this behavior, feel free to teach me some rules behind the pull request :)

I added some (in my opinion) useful features when working in a normal site environment.
- Breadcrumbs using WhiteOctobers Bundle
- Internationalisation for the templates
- Unique slug generation (bit bad how it's done but I didn't wanted to add the @Sluggable Doctrine Extension as dependencies)
- Some Console commands to update slugs and stats
- Fixed some minor bugs/missing parameters here and there
- Hierarchal Categories (parent categories)
- More SEO friendly pagination

What I did not, was to update the mongodb as I dont use them and I do not know how to markup them.

A branch would be cool, but I dont know how to create on tough (if it is even possible for outstanding ppls.)

Greez,
Blackskyliner

### CODE REVIEW

File: Command/UpdateSlugsCommand.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        /** @var $categories \Herzult\Bundle\ForumBundle\Model\Category[] */
+        $categories = $this->getContainer()->get('herzult_forum.repository.category')->findAll();
+        foreach($categories as $val)
+        {
+            $val->generateSlug();
+        }
+
+        $topics = $this->getContainer()->get('herzult_forum.repository.topic')->findAll();
```
Comment (Herzult) [inline Command/UpdateSlugsCommand.php:38]: I think it will break if you have a very big amount of topics.

File: Command/UpdateSlugsCommand.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    /**
+     * Executes the update.
+     *
+     * @param \Symfony\Component\Console\Input\InputInterface   $input  Input
+     * @param \Symfony\Component\Console\Output\OutputInterface $output Output
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        /** @var $categories \Herzult\Bundle\ForumBundle\Model\Category[] */
```
Comment (Herzult) [inline Command/UpdateSlugsCommand.php:31]: Can you remove this comments please.

File: Command/UpdateSlugsCommand.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        $categories = $this->getContainer()->get('herzult_forum.repository.category')->findAll();
+        foreach($categories as $val)
```
Comment (Herzult) [inline Command/UpdateSlugsCommand.php:33]: In Symfony coding style, it should be:

```
foreach ($categories as $category) {
    $category->generateSlug();
}
```

Else, why do we need a generateSlug() command ? can't we use any hook ?

File: Controller/CategoryController.php [hunk 4]
```diff
@@ -26,7 +34,7 @@ public function showAction($slug)
         $template = sprintf('%s:show.%s.%s', $this->container->getParameter('herzult_forum.templating.location.category'), $this->get('request')->getRequestFormat(), $this->getRenderer());
         return $this->get('templating')->renderResponse($template, array(
             'category'  => $category,
-            'page'      => $this->get('request')->query->get('page', 1)
+            'page'      => $page//$this->get('request')->query->get('page', 1)
```
Comment (Herzult) [inline Controller/CategoryController.php:37]: You should remove all the dead code, not comment it.

If some day we need to get it back, we have git ;)

File: Resources/doc/installation.markdown [hunk 5]
```diff
@@ -26,6 +26,15 @@ Update your `deps` file adding the following lines:
         git=https://github.com/l3pp4rd/DoctrineExtensions.git
         target=/doctrine-extensions
 
+    [WhiteOctoberBreadcrumbsBundle]
```
Comment (Herzult) [inline Resources/doc/installation.markdown:29]: I don't want the bundle to depend on to many bundles.

Is this bundle really needed for such simple functionality ?

File: Document/CategoryRepository.php [hunk 6]
```diff
@@ -24,6 +24,25 @@ public function findAll()
     }
 
     /**
+     * @see CategoryRepositoryInterface::findAllSubCategories
+     */
+    function findAllSubCategories($id)
+    {
+        // TODO: Implement findAllSubCategories() method.
```
Comment (Herzult) [inline Document/CategoryRepository.php:31]: Prefer `@todo` because it's a conventional comment

File: Entity/CategoryRepository.php [hunk 7]
```diff
@@ -27,6 +27,33 @@ public function findAll()
...
+    }
+
+    /**
+     * @see CategoryRepositoryInteface::findAllSubCategories
+     */
+    public function findAllSubCategories($id)
+    {
+        return $this->createQueryBuilder('c')
+                    ->select('c')
+                    ->where('c.parentCategory = ?1')
```
Comment (Herzult) [inline Entity/CategoryRepository.php:49]: Can you use named parameters instead? It's more explicit

File: Form/SearchFormType.php [hunk 8]
```diff
@@ -17,6 +17,7 @@ public function getDefaultOptions(array $options)
         return array(
             'data_class'        => 'Herzult\Bundle\ForumBundle\Search\Search',
             'csrf_protection'   => false,
+            'translation_domain' => 'HerzultForumBundle_forms',
```
Comment (Herzult) [inline Form/SearchFormType.php:20]: Maybe we should not create to many translation files, try to group all in HerzultForumBundle.xx.yml

File: Model/Topic.php [hunk 9]
```diff
@@ -64,7 +65,14 @@ public function getId()
...
         $this->subject = $subject;
-        $this->setSlug(Urlizer::urlize($this->getSubject()));
+    }
+
+    /**
+     * Generates the slug or updates it.
+     */
+    public function generateSlug()
+    {
+        $this->setSlug($this->getId()."-".$this->getSubject());
```
Comment (Herzult) [inline Model/Topic.php:75]: Can't we use "real" slugs? I mean slugs without the id

File: Command/UpdateSlugsCommand.php (hunk 1 above)
Comment (Blackskyliner) [inline Command/UpdateSlugsCommand.php:38]: Tried with 2k topics and up to 20k posts, for me it worked like a charm, don't know whats big though.
(It was real data as I wrote an importer for Joomla 1.5 Kuneai Board smthl like that, thats why I added it in the first place - In daily usage this should not be needed though.)

File: Command/UpdateSlugsCommand.php (hunk 3 above)
Comment (Blackskyliner) [inline Command/UpdateSlugsCommand.php:33]: I used it mainly to update, after importing data from another forum, as the importer was just really plain and pumped the data in the tables...

File: Controller/CategoryController.php (hunk 4 above)
Comment (Blackskyliner) [inline Controller/CategoryController.php:37]: Ohh, yea thats right :)

File: Resources/doc/installation.markdown (hunk 5 above)
Comment (Blackskyliner) [inline Resources/doc/installation.markdown:29]: Well we could emulate our own. But if someone else uses this bundle, for example, he could use it in its main-layout and spare it in the forum one and would have the full bread-crumbing. In the end we could also write our own 'lil functions which would then read this array and therefore we would have the compatibility to this additional bundle and therefore we wouldn't need to require them as dependencies. But then we would have duplicated code... Don't know which one is better though... The old Breadcrumbing was (at least for me) not suitable for adding it into some real page layout. I also hated it to correct all the templates after changing something in the breadcrubs layout. (Thats how the idea for views/Trais/breadcrumbs.html.twig started) As I wanted to have one major file to change for layouting breadcrumbs.. (After generalization I ended up having just one include in the main layout left, therefore the Traits template won't be needed anymore.

File: Model/Topic.php (hunk 9 above)
Comment (Blackskyliner) [inline Model/Topic.php:75]: The only way I would know then would be using the @Sluggable Doctrine Extension - As where do we want to define our keys?

We could use author.name-topic.subject, but what if the user posts two entries with the same name?
The ID is the only unique column we have... We could generate some uniqid() but then we could also use ID...

Even an key over the categoryid and slug or categoryname and slug wouldn't be truely unique.

Every bigger forums does an rewrite of "ID-Subject.html". To have an unique result, but I am open to any other ways, if there are any.

File: Document/CategoryRepository.php (hunk 6 above)
Comment (Blackskyliner) [inline Document/CategoryRepository.php:31]: That was more or less my IDE which I use, I will correct it though.

File: Form/SearchFormType.php (hunk 8 above)
Comment (Blackskyliner) [inline Form/SearchFormType.php:20]: Yea I also recognized it as a bad solution, but in the current project are many construction zones so I didn't correcred it till now..

File: Command/UpdateSlugsCommand.php (hunk 1 above)
Comment (Herzult) [inline Command/UpdateSlugsCommand.php:38]: Ok, that's not really a problem ;)

**Comment by Blackskyliner** [comment]
Sorry I can't get some time atm, but I will do the suggested changes soon. After I did them what to do next? add them to the pullrequest? Like I said I'm kinda new :)

**Comment by Herzult** [comment]
If you can, try to split it in many small PRs.

**Pull request closed without merging**